### PR TITLE
Add diagnostic struct for const eval error in `rustc_middle`

### PR DESCRIPTION
**Pull request by pierwill**
Part of #100717.

r? @ghost

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
I think you failed a rebase.

File: compiler/rustc_middle/src/ty/adt.rs [hunk 1]
```diff
@@ -458,11 +458,8 @@ impl<'tcx> AdtDef<'tcx> {
                     Some(Discr { val: b, ty })
                 } else {
                     info!("invalid enum discriminant: {:#?}", val);
-                    crate::mir::interpret::struct_error(
-                        tcx.at(tcx.def_span(expr_did)),
-                        "constant evaluation of enum discriminant resulted in non-integer",
-                    )
-                    .emit();
+                    let tcx = tcx.at(tcx.def_span(expr_did));
+                    tcx.sess.emit_err(crate::error::ConstEvalNonIntError { span: tcx.span });
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/adt.rs:462]:
```suggestion
                    tcx.sess.emit_err(crate::error::ConstEvalNonIntError { span: tcx.def_span(expr_did) });
```

File: src/test/ui/consts/const_in_pattern/issue-73431.stderr [hunk 2]
```diff
@@ -1 +1 @@
-WARN rustc_mir_build::thir::pattern::const_to_pat MIR const-checker found novel structural match violation. See #73448.
```
Comment (davidtwco) [inline src/test/ui/consts/const_in_pattern/issue-73431.stderr:1]: Why has this changed?

**Comment by pierwill** [comment]
I'm stilling getting strange ui test errors locally, but this passes in CI.

**Comment by compiler-errors** [comment]
@pierwill are you rebased onto the latest master? maybe that's why.

**Comment by pierwill** [comment]
> @pierwill are you rebased onto the latest master? maybe that's why.

I am. :(

**Pull request merged**